sycl/source/detail/builtins_integer.cpp

```diff
@@ -21,10 +21,23 @@ namespace d = s::detail;
 namespace __host_std {
 namespace {
 
-template <typename T> inline T __abs_diff(T x, T y) {
+template <typename T> inline std::make_unsigned_t<T> __abs_diff(T x, T y) {
   static_assert(std::is_integral<T>::value,
                 "Only integral types are supported");
-  return (x > y) ? (x - y) : (y - x);
+  if constexpr (std::is_unsigned_v<T>) {
+    return (x > y) ? (x - y) : (y - x);
+  } else {
+    // We need to be careful too avoid undefined behavior from signed integer
+    // overflow.
+    using UT = std::make_unsigned_t<T>;
+    UT ax = std::abs(x);
+    UT ay = std::abs(y);
+    // If only one of them was negative, the distance is the sum.
+    if ((x < 0) != (y < 0))
+      return ax + ay;
```

Comment:
this clause where one is positive and the other negative is the only clause  that has the risk of UB, correct?

Given these built-ins are pretty low level and need to stay performant, might it be better to only do the std::abs() conversion in this clause?  And just leave the other as  `return (x > y) ? (x - y) : (y - x);`  ?

Reply:
From a performance perspective, I would expect the compiler to be able to optimize the calls to `std::abs` given half the work is to check if the values are negative. That said, your suggestion does make for much nicer code, so I've made the changes. Thanks!